Properly handle cancelled-by-user response in HttpSource

alpaix opened the pull request:
VS credential dialog throws OperationCanceledException when user hits cancel.
Added special catch to reset authentication state of HttpSource retry loop.
After the change "cancelled" source will be blocked for following authentication attempts.

//cc @yishaigalatzer @emgarten @deepakaravindr

alpaix committed:
Properly handle cancelled-by-user response in HttpSource

VS credential dialog throws OperationCanceledException when user hits cancel.
Added special catch to reset authentication state of HttpSource retry loop.
After the change "cancelled" source will be blocked for following authentication attempts.

emgarten commented:
This looks like a good fix. 

Is there any reason the UI prompt doesn't handle this exception itself?
I like the idea of stopping after a cancel, maybe the httpsource should noop on http calls also when this is maxed and unsuccessful? All sources would keep calling it regardless, right?

pspill commented on src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs:
```diff
@@ -317,6 +317,11 @@ public Task<JObject> GetJObjectAsync(Uri uri, ILogger log, CancellationToken tok
+
+                        // null means cancelled by user
+                        // block subsequent attempts to annoy user with prompts
+                        _authRetries = HttpHandlerResourceV3Provider.MaxAuthRetries;
```
Why is _authRetries a global field of for this source, instead of a counter just for the current request?

emgarten replied:
During an install this would end up coming up constantly. For different UI actions it may make sense to retry for each. Do you have a specific scenario where a user might be unable to provide credentials and have to continue on but later have the correct password?

alpaix replied:
The motivation behind the fix was addressing a regression from 3.3:
- 3 retries of credential prompts when user hits cancel
- Failing to authenticate means restarting visual studio, given that vsts plans to deprecate credentials rather fast, I don’t think that is acceptable

alpaix committed:
Authentication ambient state

alpaix committed:
Ambient activity context

The purpose of this change is to scope source authentication state to an
abstract activity.

- Introduced an ambient context class with a single property
  'CorrelationId' identifying current activity
- HttpSource makes use of the CorrelationId by mapping it to a
  AuthenticationState, e.g. retries counter, blocking state.
- Added a call to create new activity in key places such as install
  package, uninstall package, restore missing packages, start search

alpaix commented:
//ping @rrelyea @emgarten @deepakaravindr @joelverhagen 
Please take a look as I addressed "eternal source block" issue.

alpaix commented:
Filed a https://github.com/NuGet/Home/issues/2362 describing the actual problem as reported in email.

alpaix committed:
Minor fixes

joelverhagen commented:
What's the status here?

alpaix commented:
@joelverhagen The fix wasn't approved for 3.4.xxx release. As these milestones are all behind us I'm going to merge it to dev only.

BTW, any feedback on the fix itself?

joelverhagen commented on src/NuGet.Core/NuGet.Common/ActivityCorrelationContext.cs:
```diff
@@ -0,0 +1,84 @@
+        /// <summary>
+        /// Represents activity unique ID.
+        /// </summary>
+        public string CorrelationId { get; private set; }
```
Why a string? `Guid` is smaller :).

alpaix replied:
String is generic universal representation of id that can be anything. I didn't want to limit it to a specific type. Memory footprint difference is not significant here.

joelverhagen commented on src/NuGet.Core/NuGet.Common/ActivityCorrelationContext.cs:
```diff
@@ -0,0 +1,84 @@
+#if DNXCORE50
+                return _instance.Value ?? _globalInstance;
+#else
+                var context = CallContext.LogicalGetData(DataSlotId);
```
Can we use `AsyncLocal` everywhere?

alpaix replied:
`AsyncLocal<T>` is a new class available since .NET 4.6. As the fix initially targeted 3.4.x release I had to provide alternative solution using good old `CallContext`. 

I'll go ahead and re-write it to use `AsyncLocal` everywhere for dev.

yishaigalatzer replied:
Please keep the code depending on 4.5 and not 4.6

From: Alex Panov [mailto:notifications@github.com]
Sent: Wednesday, April 13, 2016 12:58 PM
To: NuGet/NuGet.Client NuGet.Client@noreply.github.com
Cc: Yishai Galatzer yigalatz@microsoft.com
Subject: Re: [NuGet/NuGet.Client] Properly handle cancelled-by-user response in HttpSource (#388)

In src/NuGet.Core/NuGet.Common/ActivityCorrelationContext.cshttps://github.com/NuGet/NuGet.Client/pull/388#discussion_r59614494:

> - ```
>      /// <summary>
>   ```
> - ```
>      /// Represents activity unique ID.
>   ```
> - ```
>      /// </summary>
>   ```
> - ```
>      public string CorrelationId { get; private set; }
>   ```
> 
> +
> - ```
>      /// <summary>
>   ```
> - ```
>      /// Returns ambient context value or global instance if not set earlier.
>   ```
> - ```
>      /// </summary>
>   ```
> - ```
>      public static ActivityCorrelationContext Current
>   ```
> - ```
>      {
>   ```
> - ```
>          get
>   ```
> - ```
>          {
>   ```
> 
> +#if DNXCORE50
> -                return _instance.Value ?? _globalInstance;
> 
> +#else
> -                var context = CallContext.LogicalGetData(DataSlotId);

AsyncLocal<T> is a new class available since .NET 4.6. As the fix initially targeted 3.4.x release I had to provide alternative solution using good old CallContext.

I'll go ahead and re-write it to use AsyncLocal everywhere for dev.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHubhttps://github.com/NuGet/NuGet.Client/pull/388/files/796f4a979ba04ea6ca36364c77041996ff19e067#r59614494

joelverhagen commented on src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs:
```diff
@@ -29,7 +30,7 @@ public class HttpSource : IDisposable
         private readonly Uri _baseUri;
         private HttpClient _httpClient;
-        private int _authRetries;
+        private Dictionary<string, AmbientAuthenticationState> _authStates = new Dictionary<string, AmbientAuthenticationState>();
```
Can we clear this at some point?

alpaix replied:
As a possible solution we can auto-expire the state say after 1 hour. That would be similar to what we did for image download failures.

alpaix replied:
Filed https://github.com/NuGet/Home/issues/2685.

joelverhagen commented on src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs:
```diff
@@ -382,6 +391,23 @@ public Task<JObject> GetJObjectAsync(Uri uri, ILogger log, CancellationToken tok
+        {
+            var correlationId = ActivityCorrelationContext.Current.CorrelationId;
+
+            if (!_authStates.ContainsKey(correlationId))
```
nit: We can save a lookup by using `TryGetValue`, right?

alpaix replied:
Good idea! Will do

joelverhagen commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs:
```diff
@@ -74,6 +75,8 @@ protected NuGetPowerShellBaseCommand()
                 ExecutionContext = new IDEExecutionContext(_commonOperations);
             }
+
+            ActivityCorrelationContext.StartNew();
```
Could you explain the rationale behind each of these `StartNew` calls?

alpaix replied:
This creates new logical activity context associated with install, restore, search, etc

joelverhagen commented:
:watch:
